### DISCUSSION TITLE
Union support not optimal in memory usage [13725]

### DISCUSSION THREAD

**Discussion by jwillemsen** [eProsima/Fast-DDS-Gen#165] [Ideas]
The current code generation for IDL unions generates a class with for each case a plain member resulting in memory usage for all types independent of the discriminator, see for example https://github.com/eProsima/Fast-DDS/blob/6fa5e5cc8abcf506e70be203934bd3c522bd3edf/test/xtypes/idl/Types.h#L4693. This leads to a lot of memory usage when larger unions are used. This can be done much more optimized by using an union internally, see for example https://github.com/RemedyIT/idl2cpp11/blob/master/examples/union/testC.h for a possible solution

**Comment by michaelbgratton** [comment]
Is this optimization still on the table? I'm interested in this feature.  It looks to me that this could be done quite simply using an anonymous union:[https://en.cppreference.com/w/cpp/language/union]

Basically, the emitted code is already a variant (aka discriminating union), in that it has an enum type for the active member, followed by the allowed member types.  Simply surrounding these types in `union { }` should produce the desired effect.

**Comment by michaelbgratton** [comment]
To be explicit, surrounding [here](https://github.com/eProsima/Fast-DDS-Gen/blob/c020ffdb593bb53483e3e95cc79b9c506c3c5144/src/main/java/com/eprosima/fastcdr/idl/templates/TypesHeader.stg#L358) with `union{ }` should do the trick.

**Comment by jwillemsen** [comment]
It is not as simple when string or structured types are used within the union, at that moment initialization and destruction requires additional code

**Comment by michaelbgratton** [comment]
Yes, and std::vector for the sequence too.  The easier path would be to use std::variant which provides exactly the discriminating union type needed, but it can be an up-hill battle to get c++17 features accepted.

**Comment by MRicoIE2CS** [comment]
I see that this is a proposal for achieving a more optimal behavior regarding memory usage. For that reason, I proceed and convert this issue to a discussion.
Thank you all for the contribution.
